Simplification of the dbt Integration tutorial (#2371)
astronomer/docs: learn/airflow-dbt.md

---

> title: "Orchestrate dbt Core jobs with Cosmos"

Reviewer: ```suggestion
title: "Orchestrate dbt Core jobs with Airflow and Cosmos"
```
Just in case people come from google search, we should still make it clear that this is an Airflow-first guide

</head>

import CodeBlock from '@theme/CodeBlock';
import cosmos_dag from '!!raw-loader!../code-samples/dags/airflow-dbt/cosmos_dag.py';
import airflow_dbt_bashoperator from '!!raw-loader!../code-samples/dags/airflow-dbt/airflow_dbt_bashoperator.py';

[dbt Core](https://docs.getdbt.com/) is an open-source library for analytics engineering that helps users build interdependent SQL models for in-warehouse data transformation, using ephemeral compute of data warehouses. 

The open-source provider package [Cosmos](https://astronomer.github.io/astronomer-cosmos/) allows you to automatically create Airflow tasks from dbt models, seamlessly integrating dbt jobs into your Airflow orchestration environment. You can turn your dbt Core projects into an Airflow task group with just a few lines of code:

```python
DbtTaskGroup(
    group_id="dbt_project_1",
    dbt_project_name=DBT_PROJECT_NAME,
    conn_id=CONNECTION_ID,
    dbt_root_path=DBT_ROOT_PATH,
    dbt_args={
        "dbt_executable_path": DBT_EXECUTABLE_PATH,
        "schema": SCHEMA_NAME,
        "vars": '{"my_key": "my_value"}',
    },
)
```

:::info

> If you are already familiar with Airflow and dbt Core check out this [github repository](https://github.com/astronomer/astro-dbt-provider-tutorial-example).

Reviewer: This doesn't explain what the repo is or why I would want to check it out. Maybe reword to be "If you are familiar with Airflow and dbt Core and just want to get your project running, copy the example project from [Astronomer GitHub](https://github.com/astronomer/astro-dbt-provider-tutorial-example) and run it locally using the Astro CLI."

Author: Yeah, good point. does this work:

> If you are already familiar with Airflow and dbt Core and just want to get a project running, clone [Astronomers Cosmos example repository](https://github.com/astronomer/astro-dbt-provider-tutorial-example) and run it locally using the Astro CLI. 

? :)

:::

> :::info
> 

Reviewer: ```suggestion
```

> For a tutorial on how to use dbt Cloud with Airflow see [Orchestrate dbt Cloud with Airflow](airflow-dbt-cloud.md).

Reviewer: ```suggestion
For a tutorial on how to use dbt Cloud with Airflow, see [Orchestrate dbt Cloud with Airflow](airflow-dbt-cloud.md).
```

> :::
> 

Reviewer: ```suggestion
```

## Why use Airflow with dbt Core?

dbt Core offers the possibility to build modular, reuseable SQL components with built-in dependency management and [incremental builds](https://docs.getdbt.com/docs/build/incremental-models). With [Cosmos](https://astronomer.github.io/astronomer-cosmos/) you can integrate dbt jobs into your Airflow orchestration environment as a standalone DAG or as a task group within a DAG. 

The benefits of using Airflow with dbt Core include:

> - Use Airflow's [data-aware scheduling](airflow-datasets.md) and [Airflow sensors](what-is-a-sensor.md) to run models depending on any other event in your data ecosystem.

Reviewer: ```suggestion
- Use Airflow's [data-aware scheduling](airflow-datasets.md) and [Airflow sensors](what-is-a-sensor.md) to run models depending on other events in your data ecosystem.
```

- Turn each dbt model into a task, complete with Airflow features like [retries](rerunning-dags.md#automatically-retry-tasks) and [error notifications](error-notifications-in-airflow.md), as well as full observability into past runs directly in the Airflow UI.
- Run `dbt test` on tables created by individual models immediately after a model has completed. Catch issues before moving downstream and integrate additional [data quality checks](data-quality.md) with your preferred tool to run alongside dbt tests.
- Run dbt projects using [Airflow connections](connections.md) instead of dbt profiles. You can store all your connections in one place, directly within Airflow or by using a [secrets backend](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/secrets-backend/index.html).
- Leverage native support for installing and running dbt in a virtual environment to avoid dependency conflicts with Airflow.

## Time to complete

> This tutorial takes approximately 30 minutes to complete.

Reviewer: Nice 😎

- Airflow operators. See [Operators 101](what-is-an-operator.md).
- Airflow task groups. See [Airflow task groups](task-groups.md).
- Airflow connections. See [Manage connections in Apache Airflow](connections.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview).
- Access to a data warehouse supported by dbt Core. See [dbt documentation](https://docs.getdbt.com/docs/supported-data-platforms) for all supported warehouses. This tutorial uses a local [PostgreSQL](https://www.postgresql.org/) database.
- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview) with an empty Astro project created by `astro dev init`.
- Access to a data warehouse supported by dbt Core. See [dbt documentation](https://docs.getdbt.com/docs/supported-data-platforms) for all supported warehouses. This tutorial uses a Postgres database.

You do not need to have dbt Core installed locally in order to complete this tutorial.

## Step 1: Configure your Astro project

An Astro project contains all of the files you need to run Airflow locally.
To use dbt with Airflow install dbt Core in a virtual environment and the Astro dbt provider in your Astro project.

> 1. Create a new Astro project:

Reviewer 2: We should leave this in. Generally we want to simplify the example DAGs shown in these tutorials to (wherever possible) only use the tool in question. But, we still need to provide context around setting up the project, etc.

    RUN python -m venv dbt_venv && source dbt_venv/bin/activate && \
    pip install --no-cache-dir dbt-postgres && deactivate
    ```

    This code runs a bash command when the Docker image is built that creates a virtual environment called `dbt_venv` inside of the Astro CLI scheduler container. The `dbt-postgres` package, which also contains `dbt-core`, is installed in the virtual environment. If you are using a different data warehouse, replace `dbt-postgres` with the adapter package for your data warehouse.

2. Add the [Astro dbt provider package](https://github.com/astronomer/astronomer-cosmos) to your Astro project `requirements.txt` file.

    ```text
    astronomer-cosmos==0.6.8
    ```

## Step 2: Prepare your dbt project

> In order to integrate your dbt project with Airflow, you need to add the dbt project in a folder called `dbt` in the `dags` folder. You can either add your own project or follow the steps below to create a simple project using two models.

Reviewer 2: ```suggestion
In order to integrate your dbt project with Airflow, you need to add your dbt project in a subfolder called `dbt` within your `dags` folder. You can either add your own project, or follow the steps below to create a simple project using two models.
```

2. In the `dbt` folder, create a folder called `my_simple_dbt_project`.

3. In the `my_simple_dbt_project` folder add your `dbt_project.yml`. This configuration file needs to contain at least the name of the project. This tutorial additionally shows how to inject a variable called `my_name` from Airflow into your dbt project.

    ```yaml
    name: 'my_simple_dbt_project'
    vars:
        my_name: "No entry"
    ```

4. Add your dbt models in a folder called `models` within the `my_simple_dbt_project` folder. You can add as many models as you want to run. This tutorial uses the following two models:

    `model1.sql`:

    ```sql
    SELECT '{{ var("my_name") }}' as name
    ```

    `model2.sql`:

    ```sql
    SELECT * FROM {{ ref('model1') }}
    ```

    `model1.sql` selects the variable `my_name`. `model2.sql` depends on `model1.sql` and selects everything from the upstream model.

You should now have the following structure within your Astro project:

```text
.
└── dags
    └── dbt
        └── my_simple_dbt_project
           ├── dbt_project.yml
           └── models
               ├── model1.sql
               └── model2.sql
```

> ## Step 3: Create an Airflow connection to your data warehouse

Reviewer 2: I think totally fine to leave this in here even if we have a connections guide. As I mentioned above, we still want to provide the setup and context throughout the tutorial, and want to limit how much people need to click into other docs to find the info they need.

    $ astro dev init
    ```

2. Open the `Dockerfile` and add the following lines to the end of the file:

    ```text
    # replace dbt-postgres with another supported adapter if you're using a different warehouse type
    RUN python -m venv dbt_venv && source dbt_venv/bin/activate && \
    pip install --no-cache-dir dbt-postgres && deactivate
    ```

    This code runs a bash command when the Docker image is built that creates a virtual environment called `dbt_venv` inside of the Astro CLI scheduler container. The `dbt-postgres` package, which also contains `dbt-core`, is installed in the virtual environment. If you are using a different data warehouse, replace `dbt-postgres` with the adapter package for your data warehouse.

3. Add [Cosmos](https://github.com/astronomer/astronomer-cosmos) to your Astro project `requirements.txt` file.

    ```text
    astronomer-cosmos==0.6.8
    ```

## Step 2: Prepare your dbt project

> In order to integrate your dbt project with Airflow, you need to add your dbt project in a subfolder called `dbt` within your `dags` folder. You can either add your own project, or follow the steps below to create a simple project using two models.

Reviewer: ```suggestion
To integrate your dbt project with Airflow, you need to add your dbt project to a subfolder called `dbt` within your `dags` folder. You can either add your own project, or follow the steps below to create a simple project using two models.
```

1. Create a folder called `dbt` in your `dags` folder. 

2. In the `dbt` folder, create a folder called `my_simple_dbt_project`.

3. In the `my_simple_dbt_project` folder add your `dbt_project.yml`. This configuration file needs to contain at least the name of the project. This tutorial additionally shows how to inject a variable called `my_name` from Airflow into your dbt project.

    ```yaml
    name: 'my_simple_dbt_project'
    vars:
        my_name: "No entry"
    ```

> 4. Add your dbt models in a folder called `models` within the `my_simple_dbt_project` folder. You can add as many models as you want to run. This tutorial uses the following two models:

Reviewer: ```suggestion
4. Add your dbt models in a subfolder called `models` in the `my_simple_dbt_project` folder. You can add as many models as you want to run. This tutorial uses the following two models:
```

    `model1.sql`:

    ```sql
    SELECT '{{ var("my_name") }}' as name
    ```

    `model2.sql`:

    ```sql
    SELECT * FROM {{ ref('model1') }}
    ```

    `model1.sql` selects the variable `my_name`. `model2.sql` depends on `model1.sql` and selects everything from the upstream model.

You should now have the following structure within your Astro project:

```text
.
└── dags
    └── dbt
        └── my_simple_dbt_project
           ├── dbt_project.yml
           └── models
               ├── model1.sql
               └── model2.sql
```

## Step 3: Create an Airflow connection to your data warehouse

> Cosmos allows you to use Airflow connections inside your dbt project. 

Reviewer: ```suggestion
Cosmos allows you to apply Airflow connections to your dbt project. 
```
How would you connect to the database from a non-integrated dbt project? Does Airflow make that process easier? We could mention that if so.

Author: If you would be running dbt by itself, not with Airflow? You would define the connection in a profiles.yaml file: https://docs.getdbt.com/docs/core/connect-data-platform/connection-profiles 
The info that has to be provided is the same, so not necessarily easier, the upside of using Cosmos is to be able to hand the information from the Airflow connection to the dbt project, so you only need to define and maintain the connection in one place.

:::

## Step 4: Write your Airflow DAG

> The DAG you'll write uses the Astro dbt provider to create tasks from existing dbt models. You can add up- and downstream tasks to and embeds the dbt project within other actions in your data ecosystem.

Reviewer 2: ```suggestion
The DAG you'll write uses the Astro dbt provider to create tasks from existing dbt models. You can also add up- and downstream tasks to embed the dbt project within other actions in your data ecosystem.
```
Double check me on this edit - I _think_ this is what you were trying to say?

    <CodeBlock language="python">{cosmos_dag}</CodeBlock>

    This DAG uses the `DbtTaskGroup` class from the Astro dbt provider to create a task group from the models in your dbt project. Dependencies between your dbt models are automatically turned into dependencies between Airflow tasks.
    Using the `vars` keyword in the dictionary provided to the `dbt_args` parameter, you can inject variables into the dbt project. This DAG injects `Astro` for the `my_name` variable. If your dbt project contains dbt tests, they will be run directly after a model has completed.

3. Run the DAG manually by clicking the play button and view the DAG in the graph view. Double click the task groups in order to expand them and see all tasks. 

    ![Cosmos DAG graph view](/img/guides/cosmos_dag_graph_view.png)

:::info

The DbtTaskGroup class populates an Airflow task group with Airflow tasks created from dbt models inside of a normal DAG. To directly define a full DAG containing only dbt models use the `DbtDag` class, as shown in the [Astro dbt provider documentation](https://astronomer.github.io/astronomer-cosmos/dbt/usage.html#full-dag).

:::

> ## Alternative ways to run dbt Core with Airflow

Reviewer 2: Do we think it would make sense to put this section below the Conclusion? Obviously "Conclusion" seems to make the most sense at the end, but it seems a little weird to talk about alternatives and then go back to say congrats for using Cosmos... or maybe we we title "conclusion" something else...

Author: Hmmm... yeah, I think back when we start adding the conclusion we barely had tutorials with appendix information like this. I vote for putting it before the alternative ways.... it feels more natural to me than expanding it to say "we also congratulate you for learning xyz" 😄 

How about calling it "Next steps" ? (I feel a slight deja vu, sorry if we discussed that before)

:::

## Step 4: Write your Airflow DAG

> The DAG you'll write uses Cosmos to create tasks from existing dbt models. You can add up- and downstream tasks to embed the dbt project within other actions in your data ecosystem.

Reviewer: ```suggestion
The DAG you'll write uses Cosmos to create tasks from existing dbt models. You can add upstream and downstream tasks to embed the dbt project within other actions in your data ecosystem.
```

2. Copy and paste the following DAG code into the file:

    <CodeBlock language="python">{cosmos_dag}</CodeBlock>

    This DAG uses the `DbtTaskGroup` class from the Cosmos package to create a task group from the models in your dbt project. Dependencies between your dbt models are automatically turned into dependencies between Airflow tasks.
    Using the `vars` keyword in the dictionary provided to the `dbt_args` parameter, you can inject variables into the dbt project. This DAG injects `Astro` for the `my_name` variable. If your dbt project contains dbt tests, they will be run directly after a model has completed.

3. Run the DAG manually by clicking the play button and view the DAG in the graph view. Double click the task groups in order to expand them and see all tasks. 

    ![Cosmos DAG graph view](/img/guides/cosmos_dag_graph_view.png)

:::info

The DbtTaskGroup class populates an Airflow task group with Airflow tasks created from dbt models inside of a normal DAG. To directly define a full DAG containing only dbt models use the `DbtDag` class, as shown in the [Cosmos documentation](https://astronomer.github.io/astronomer-cosmos/dbt/usage.html#full-dag).

:::

> ## Next steps

Reviewer: ```suggestion
```

Reviewer: I don't think we need a header if there's not an actual "Next step" to take here
